File: problems/100_same_tree.py
from typing import Self


class TreeNode:
    def __init__(self, val: int | None, left: Self | None, right: Self | None):
        self.val = val
        self.left = left
        self.right = right


# Iterative solution.
class Solution:
    def isSameTree(self, p: TreeNode | None, q: TreeNode | None) -> bool:
        stack1 = [p]
        stack2 = [q]

        while stack1:
            n1 = stack1.pop()
            n2 = stack2.pop()

            if not n1 and not n2:
                continue

            n1_val = n1.val if n1 else None
            n2_val = n2.val if n2 else None

            if n1_val != n2_val:
                return False

            if n1:
                stack1.append(n1.left)
                stack1.append(n1.right)

            if n2:
                stack2.append(n2.left)
                stack2.append(n2.right)

        if len(stack1) != len(stack2):
            return False

        return True


# Recursive solution.
class RecursiveSolution:
    def isSameTree(self, p: TreeNode | None, q: TreeNode | None) -> bool:
        if not p and not q:
            return True

        p_val = p.val if p else None
        q_val = q.val if q else None

        if p_val != q_val:
            return False

        p_left = p.left if p else None
        p_right = p.right if p else None
        q_left = q.left if q else None
        q_right = q.right if q else None

        return self.isSameTree(p_left, q_left) and self.isSameTree(p_right, q_right)

File: problems/100-same-tree.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(
    val?: number,
    left: TreeNode | null = null,
    right: TreeNode | null = null
  ) {
    this.val = val ?? 0
    this.left = left
    this.right = right
  }
}

// Recursive solution.
function isSameTreeRecursive(p: TreeNode | null, q: TreeNode | null) {
  if (!p && !q) {
    return true
  } else if (!q || !p) {
    return false
  } else {
    return (
      p.val === q.val &&
      isSameTree(p.left, q.left) &&
      isSameTree(p.right, q.right)
    )
  }
}

// Iterative solution.
function isSameTree(p: TreeNode | null, q: TreeNode | null) {
  const pStack = [p]
  const qStack = [q]

  while (pStack.length) {
    const n1 = pStack.pop()
    const n2 = qStack.pop()
    if (n1?.val !== n2?.val) {
      return false
    }
    if (n1) {
      pStack.push(n1.left)
      pStack.push(n1.right)
    }
    if (n2) {
      qStack.push(n2.left)
      qStack.push(n2.right)
    }
  }

  if (qStack.length) {
    return false
  }

  return true
}

export {}
